suppressPackageStartupMessages(library("gridExtra"))
suppressPackageStartupMessages(library('tidyverse'))
suppressPackageStartupMessages(library("dplyr"))


snp_annotate <- function(vcffile, gfffile, prefix) {
    ##calculate how many snps are in annotated regions regions

    vcf=read_tsv(vcffile, comment='##') %>%
        filter(FILTER=='PASS')

    names=colnames(vcf)[10:dim(vcf)[2]]
    for (i in 1:length(names)){
        newname=gsub('.fasta', '',names[i])
        names[i]=newname
    }
    colnames(vcf)[10:dim(vcf)[2]]=names

    qcol=match(prefix, colnames(vcf))
    snplist=vcf[vcf[,qcol]==1,]

    gff=read_tsv(gfffile, comment='#', col_names=FALSE)
    colnames(gff)=c('contig', 'caller', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'gene')

    inregion=adply(snplist, 1, function(x) {
        tiggenes=gff[gff$contig==x[1],]
        if(dim(tiggenes)[1]>0){
            ingene=sum(x[2]>tiggenes$start && x[2]<tiggenes$end)
            x$ingene=ingene
        }else{
            x$ingene=0
        }
        return(x)
    })
    return(sum(inregion$ingene))
}

snp_filter  <- function(reportfile, mincov, minratio, sampname) {
    ##take in snp report csv generated by snps_report.py
    cols=c('chr','pos','alt','ref','ref_count','alt_count','type','start','end','prod','gene','note')
    vcf=read_csv(reportfile, col_names=cols) %>%
        mutate(altratio=alt_count/(alt_count+ref_count)) %>%
        filter(alt_count>mincov) %>%
        filter(altratio>minratio) %>%
        mutate(region_size=end-start) %>%
        group_by(chr, pos) %>%
        top_n(-1, region_size) %>%
        group_by(chr, pos) %>%
        filter(row_number()==n()) %>%
        mutate(sampname=sampname)
    return(vcf)
}

snpeff_filter <- function(reportfile, mincov, sampname) {
    vcf=read_csv(reportfile) %>%
        mutate(altratio=ao/(ao+ro)) %>%
        mutate(cov=ao+ro) %>%
        mutate(samp=sampname) %>%
        filter(cov>mincov)
    return(vcf)
}
    



##coding region of 178, test for all annotations
vcf178='~/Dropbox/yfan/hardwick/parsnp_long/strain_long_snps.vcf'
gff178="/kyber/Data/NGS/projects/190513_hardwick/augustus/178_all.gff"
prefix="197_over20k"

snp_annotate(vcf178, gff178, prefix)

repfile178=paste0(datadir, 'vars/178.csv')
repfile197=paste0(datadir, 'vars/197.csv')
repfile1694=paste0(datadir, 'vars/1694.csv')
repfile6341=paste0(datadir, 'vars/6341.csv')


rep178=snp_filter(repfile178, 15, .8, '178')
rep197=snp_filter(repfile197, 15, .8, '197')
rep1694=snp_filter(repfile1694, 15, .8, '1694')
rep6341=snp_filter(repfile6341, 15, .8, '6341')
all=rbind(rep178, rep197, rep1694, rep6341)


##check out data from snp calling
datadir='/kyber/Data/NGS/projects/190513_hardwick/'
dbxdir='~/Dropbox/yfan/hardwick/'

qcfile=paste0(dbxdir, 'plots/vars_qc.pdf')
pdf(qcfile, h=5, w=12)
qcplot=ggplot(all, aes(x=altratio, colour=sampname, fill=sampname, alpha=.3)) +
    geom_density() +
    ggtitle('alt allele ratio') +
    theme_bw()
print(qcplot)
##okay, looks like reads most agree internally for all samps
dev.off()


##find which snps have which samps
snpsamps=all %>%
    group_by(chr, pos, alt, ref, type, start, end, prod, gene, note, region_size) %>%
    summarise(s178='178' %in% sampname,
              s197='197' %in% sampname,
              s1684='1694' %in% sampname,
              s6341='6341' %in% sampname)

sampnames=c('s178', 's197', 's1684', 's6341')


labsnps=snpsamps %>%
    filter(s178!=s197) %>%
    filter(type!='region') %>%
    filter(prod!='no product info') %>%
    filter(prod!='product=hypothetical protein')





##snpeff data
repfile178=paste0(datadir, 'vars/178.snpeff.csv')
repfile197=paste0(datadir, 'vars/197.snpeff.csv')
repfile1694=paste0(datadir, 'vars/1694.snpeff.csv')
repfile6341=paste0(datadir, 'vars/6341.snpeff.csv')

snpeff178=snpeff_filter(repfile178, 15, '178')
snpeff197=snpeff_filter(repfile197, 15, '197')
snpeff1694=snpeff_filter(repfile1694, 15, '1694')
snpeff6341=snpeff_filter(repfile6341, 15, '6341')

all=rbind(snpeff178, snpeff197, snpeff1694, snpeff6341)

##number of vars relative to ref
allbysamp=all %>%
    group_by(chrm, pos, ref, samp) %>%
    filter(row_number()==1)
numvars=tibble(samps=c('178', '197', '1694', '6341'))
numvars=numvars %>%
    rowwise() %>%
    mutate(numsnps=sum(allbysamp$samp==samps))
write_csv(numvars, paste0(dbxdir, 'numvars_vs_reference.csv'))


##plot ratio of alt reads to total cov
allbyvars=all %>%
    group_by(chrm, pos, samp) %>%
    filter(row_number()==n())

outfile=paste0(dbxdir, 'plots/callratios.pdf')
pdf(outfile, w=15, h=11)
plot=ggplot(allbyvars, aes(x=altratio, colour=samp, fill=samp, alpha=.20)) +
    geom_density() +
    ggtitle('Reference vs alt snp ratios') +
    xlab('Ratio') +
    theme_bw()
print(plot)
dev.off()
    
##find unique vars between samps
allsnps=all %>%
    filter(altratio>.75)
snpsdf=allsnps %>%
    group_by(chrm, pos , ref, alt) %>%
    summarise(s178='178' %in% samp,
              s197='197' %in% samp,
              s1694='1694' %in% samp,
              s6341='6341' %in% samp)
sampnames=c('s178', 's197', 's1694', 's6341')
confusion=matrix(nrow=4, ncol=4)
rownames(confusion)=sampnames
colnames(confusion)=sampnames
for (i in sampnames) {
    for (j in sampnames) {
        diffs=sum(snpsdf[,i]!=snpsdf[,j])
        confusion[i,j]=diffs
    }
}

