suppressPackageStartupMessages(library("gridExtra"))
suppressPackageStartupMessages(library('tidyverse'))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("stringdist"))


snp_annotate <- function(vcffile, gfffile, prefix) {
    ##calculate how many snps are in annotated regions regions

    vcf=read_tsv(vcffile, comment='##') %>%
        filter(FILTER=='PASS')

    names=colnames(vcf)[10:dim(vcf)[2]]
    for (i in 1:length(names)){
        newname=gsub('.fasta', '',names[i])
        names[i]=newname
    }
    colnames(vcf)[10:dim(vcf)[2]]=names

    qcol=match(prefix, colnames(vcf))
    snplist=vcf[vcf[,qcol]==1,]

    gff=read_tsv(gfffile, comment='#', col_names=FALSE)
    colnames(gff)=c('contig', 'caller', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'gene')

    inregion=adply(snplist, 1, function(x) {
        tiggenes=gff[gff$contig==x[1],]
        if(dim(tiggenes)[1]>0){
            ingene=sum(x[2]>tiggenes$start && x[2]<tiggenes$end)
            x$ingene=ingene
        }else{
            x$ingene=0
        }
        return(x)
    })
    return(sum(inregion$ingene))
}

snp_filter  <- function(reportfile, mincov, minratio, sampname) {
    ##take in snp report csv generated by snps_report.py
    cols=c('chr','pos','alt','ref','ref_count','alt_count','type','start','end','prod','gene','note')
    vcf=read_csv(reportfile, col_names=cols) %>%
        mutate(altratio=alt_count/(alt_count+ref_count)) %>%
        filter(alt_count>mincov) %>%
        filter(altratio>minratio) %>%
        mutate(region_size=end-start) %>%
        group_by(chr, pos) %>%
        top_n(-1, region_size) %>%
        group_by(chr, pos) %>%
        filter(row_number()==n()) %>%
        mutate(sampname=sampname)
    return(vcf)
}


snpeff_filter <- function(reportfile, sampname) {
    vcf=read_csv(reportfile) %>%
        mutate(altratio=ao/(ao+ro)) %>%
        mutate(cov=ao+ro) %>%
        mutate(samp=sampname)
    return(vcf)
}
    

 
##coding region of 178, test for all annotations
vcf178='~/Dropbox/yfan/hardwick/parsnp_long/strain_long_snps.vcf'
gff178="/kyber/Data/NGS/projects/190513_hardwick/augustus/178_all.gff"
prefix="197_over20k"

snp_annotate(vcf178, gff178, prefix)

repfile178=paste0(datadir, 'vars/178.csv')
repfile197=paste0(datadir, 'vars/197.csv')
repfile1694=paste0(datadir, 'vars/1694.csv')
repfile6341=paste0(datadir, 'vars/6341.csv')


rep178=snp_filter(repfile178, 15, .8, '178')
rep197=snp_filter(repfile197, 15, .8, '197')
rep1694=snp_filter(repfile1694, 15, .8, '1694')
rep6341=snp_filter(repfile6341, 15, .8, '6341')
all=rbind(rep178, rep197, rep1694, rep6341)


##find which snps have which samps
snpsamps=all %>%
    group_by(chr, pos, alt, ref, type, start, end, prod, gene, note, region_size) %>%
    summarise(s178='178' %in% sampname,
              s197='197' %in% sampname,
              s1684='1694' %in% sampname,
              s6341='6341' %in% sampname)

sampnames=c('s178', 's197', 's1684', 's6341')



##check out data from snp calling
datadir='/kyber/Data/NGS/projects/190513_hardwick/'
dbxdir='~/Dropbox/yfan/hardwick/'

qcfile=paste0(dbxdir, 'plots/vars_qc.pdf')
pdf(qcfile, h=5, w=12)
qcplot=ggplot(all, aes(x=altratio, colour=sampname, fill=sampname, alpha=.3)) +
    geom_density() +
    ggtitle('alt allele ratio') +
    theme_bw()
print(qcplot)
##okay, looks like reads most agree internally for all samps
dev.off()



labsnps=snpsamps %>%
    filter(s178!=s197) %>%
    filter(type!='region') %>%
    filter(prod!='no product info') %>%
    filter(prod!='product=hypothetical protein')





##snpeff data
repfile178=paste0(datadir, 'vars/178.snpeff.csv')
repfile197=paste0(datadir, 'vars/197.snpeff.csv')
repfile1694=paste0(datadir, 'vars/1694.snpeff.csv')
repfile6341=paste0(datadir, 'vars/6341.snpeff.csv')

snpeff178=snpeff_filter(repfile178, '178')
snpeff197=snpeff_filter(repfile197, '197')
snpeff1694=snpeff_filter(repfile1694, '1694')
snpeff6341=snpeff_filter(repfile6341, '6341')

all=rbind(snpeff178, snpeff197, snpeff1694, snpeff6341)
allsnps=all %>%
    filter(altratio>.5) %>% 
    filter(cov>15)


##number of vars relative to ref
'''
allbysamp=allsnps %>%
    group_by(chrm, pos, ref, alt, ro, ao, prot_change, altratio, cov, samp) %>%
    summarise(annot=str_c(annotation, collapse='|'), feat_type=str_c(feature_type, collapse='|'))
'''
allbysamp=allsnps %>%
    group_by(chrm, pos, ref, samp) %>%
    filter(row_number()==1)

numvars=tibble(samps=c('178', '197', '1694', '6341'))
numvars=numvars %>%
    rowwise() %>%
    mutate(numsnps=sum(allbysamp$samp==samps))
write_csv(numvars, paste0(dbxdir, 'numvars_vs_reference.csv'))




##plot ratio of alt reads to total cov
allbyvars=all %>%
    group_by(chrm, pos, samp) %>%
    filter(row_number()==n())

outfile=paste0(dbxdir, 'plots/callratios.pdf')
pdf(outfile, w=15, h=8)
plot=ggplot(allbyvars, aes(x=altratio, colour=samp, fill=samp, alpha=.20)) +
    geom_density() +
    ggtitle('Reference vs alt snp ratios') +
    xlab('Ratio') +
    theme_bw()
print(plot)
dev.off()




##find unique vars between samps
snpsdf=all %>%
    group_by(chrm, pos , ref, alt) %>%
    summarise(s178='178' %in% samp,
              s197='197' %in% samp,
              s1694='1694' %in% samp,
              s6341='6341' %in% samp)

collapse_snp_pair <- function(snppair) {
    ##same variant often appears in two samps, but labeled with different len ref/alt
    ##find these weirdly labeled ones and collapse them to the same position
    ##group by chrm/pos to find the final tally of how different the two var sets are
    clustpair=snppair %>%
        arrange(chrm, pos) %>%
        group_by(chrm) %>%
        mutate(origpos=pos) %>%
        mutate(diffpos=c(diff(pos),0)) %>%
        group_by(chrm, pos) %>%
        mutate(truediff=max(diffpos)) %>%
        mutate(toadd=any(truediff<=max(str_length(ref),str_length(alt)))) %>%
        mutate(newpos=case_when(toadd ~ pos+truediff, T ~ pos)) %>%
        ungroup() %>%
        mutate(pos=newpos)
    change=TRUE
    iter=0
    while (change==TRUE) {
        iter=iter+1
        print(iter)
        clustpairnew=clustpair %>%
            arrange(chrm, pos) %>%
            group_by(chrm) %>%
            mutate(diffpos=c(diff(pos),0)) %>%
            group_by(chrm, pos) %>%
            mutate(truediff=max(diffpos)) %>%
            mutate(toadd=any(truediff<=max(str_length(ref), str_length(alt)))) %>%
            mutate(newpos=case_when(toadd ~ pos+truediff, T ~ pos)) %>%
            ungroup() %>%
            mutate(pos=newpos)
        if (identical(clustpairnew,clustpair)) {
            change=FALSE
        }
        clustpair=clustpairnew
    }
    return(clustpairnew)
}

check_substring <- function(groupdf) {
    ##alts=grepl(groupdf$alt[1], groupdf$alt, fixed=TRUE)
    if (dim(groupdf)[1]>0) {
        dists=stringdist(groupdf$alt[1], groupdf$alt, method='lcs')
        totallen=nchar(groupdf$alt)+nchar(groupdf$alt[1])
        chars=nchar(groupdf$alt)
        allchars=matrix(nrow=length(chars), ncol=length(chars))
        for (i in 1:length(chars)){
            for (j in 1:length(chars)) {
            allchars[i,j]=min(chars[i], chars[j])
            }
        }
        
        for (i in groupdf$alt[-1]) {
            ##alts=rbind(alts, grepl(i, groupdf$alt, fixed=TRUE))
            dists=rbind(dists, stringdist(i, groupdf$alt, method='lcs'))
            totallen=rbind(totallen, nchar(groupdf$alt)+nchar(i))
        }
        
        overlaps=(totallen-dists)/2/allchars>.5
        if ( length(overlaps)>1) {
            hits=rowSums(overlaps)
        }else{
            hits=1
        }
        groupdf_sums=groupdf %>%
            mutate(subsum=hits)
    }else{
        groupdf_sums=groupdf
    }
    return(groupdf_sums)
}
                     

sampnames=c('s178', 's197', 's1694', 's6341')
confusion=matrix(nrow=4, ncol=4)
rownames(confusion)=sampnames
colnames(confusion)=sampnames
for (i in sampnames) {
    for (j in sampnames) {
        if (is.na(confusion[i,j]) & i!=j) {
            print(paste0(i, ' ', j))
            diffs=snpsdf[snpsdf[,i]!=snpsdf[,j],]
            collapsed_diffs=collapse_snp_pair(diffs)
            ##each group as two rows max
            collapsed=collapsed_diffs %>%
                group_by(chrm, pos) %>%
                group_modify(~ check_substring(.x)) %>%
                ##filter(subsum==1 && revsubsum==1) %>%
                filter(subsum<=1) %>%
                group_by(chrm, origpos) %>%
                group_modify(~ check_substring(.x)) %>%
                ##filter(subsum==1 && revsubsum==1) %>%
                filter(subsum<=1) %>%
                ungroup() %>%
                select(-pos) %>%
                mutate(pos=origpos) %>%
                select(-origpos)
            
            ##now filter for frequency, etc
            isamp=substr(i,2,5)
            jsamp=substr(j, 2,5)
            collapsefreq=all %>%
                inner_join(collapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
                filter(samp==isamp | samp==jsamp) %>%
                group_by(chrm, pos, ref, alt) %>%
                filter(row_number()==n()) %>%
                filter(cov>15 & altratio>.5) %>%
                arrange(chrm, pos)
            confusion[i,j]=dim(collapsefreq)[1]
            confusion[j,i]=dim(collapsefreq)[1]
        }
    }
}

write_csv(data.frame(confusion), paste0(dbxdir, 'snp_diffs.csv'))



##vars between 197 and 178
ptsnpsdf=snpsdf %>%
    filter(s197!=s178)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=allbysamp %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='197' | samp=='178') %>%
    group_by(chrm, pos, ref, alt) %>%
    filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos)
write_csv(ptcollapsefreq, paste0(dbxdir, '197vs178.csv'))


##vars between 197 and 1694
ptsnpsdf=snpsdf %>%
    filter(s197!=s1694)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=all %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='197' | samp=='1694') %>%
    ##group_by(chrm, pos, ref, alt) %>%
    ##filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos) %>%
    ##filter(impact!='MODIFIER') %>%
    filter(impact=='HIGH') %>%
    filter(prot_change!='')
write_csv(ptcollapsefreq, paste0(dbxdir, '197vs1694.csv'))


##vars between 197 and 6341
ptsnpsdf=snpsdf %>%
    filter(s197!=s6341)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=all %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='197' | samp=='6341') %>%
    ##group_by(chrm, pos, ref, alt) %>%
    ##filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos) %>%
    ##filter(impact!='MODIFIER') %>%
    filter(impact=='HIGH') %>%
    filter(prot_change!='')
write_csv(ptcollapsefreq, paste0(dbxdir, '197vs6341.csv'))



##vars between 178 and 1694
ptsnpsdf=snpsdf %>%
    filter(s178!=s1694)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=all %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='178' | samp=='1694') %>%
    ##group_by(chrm, pos, ref, alt) %>%
    ##filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos) %>%
    ##filter(impact!='MODIFIER') %>%
    filter(impact=='HIGH') %>%
    filter(prot_change!='')
write_csv(ptcollapsefreq, paste0(dbxdir, '178vs1694.csv'))


##vars between 178 and 6341
ptsnpsdf=snpsdf %>%
    filter(s178!=s6341)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=all %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='178' | samp=='6341') %>%
    ##group_by(chrm, pos, ref, alt) %>%
    ##filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos) %>%
    ##filter(impact!='MODIFIER') %>%
    filter(impact=='HIGH') %>%
    filter(prot_change!='')
write_csv(ptcollapsefreq, paste0(dbxdir, '178vs6341.csv'))

##vars between 1694 and 6341
ptsnpsdf=snpsdf %>%
    filter(s1694!=s6341)
ptdiffs=collapse_snp_pair(ptsnpsdf)
ptcollapsed=ptdiffs %>%
    group_by(chrm, pos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    group_by(chrm, origpos) %>%
    group_modify(~ check_substring(.x)) %>%
    ##filter(subsum==1 && revsubsum==1) %>%
    filter(subsum<=1) %>%
    ungroup() %>%
    select(-pos) %>%
    mutate(pos=origpos) %>%
    select(-origpos)
ptcollapsefreq=all %>%
    inner_join(ptcollapsed, by=c('chrm', 'pos', 'ref','alt')) %>%
    filter(samp=='1694' | samp=='6341') %>%
    ##group_by(chrm, pos, ref, alt) %>%
    ##filter(row_number()==n()) %>%
    filter(cov>15 & altratio>.5) %>%
    arrange(chrm, pos) %>%
    ##filter(impact!='MODIFIER') %>%
    filter(impact=='HIGH') %>%
    filter(prot_change!='')
write_csv(ptcollapsefreq, paste0(dbxdir, '1694vs6341.csv'))




